### PR TITLE
[DOCS] Document shard sizing guide

### LINKED ISSUE #52541: Add a documentation guide on ES shard sizing
The most popular guide available today on ES shard sizing is on the Elastic blog: https://www.elastic.co/blog/how-many-shards-should-i-have-in-my-elasticsearch-cluster

Unfortunately, the blog entry above was created in 2017 and is not actively maintained. Let's move this content over to our documentation and determine what information (if any) should be updated within the guide. 

By moving this content to our documentation, it will be easier for the team to maintain the recommendations and make changes when needed.

### LINKED ISSUE #61306: Document total_shards_per_node as a recipe for hotspots
`index.routing.allocation.total_shards_per_node` has become a very common recommendation for dealing with hotspots in a cluster that has suboptimal shard allocation.  For example, [shards of the most actively written indices accumulating](https://github.com/elastic/elasticsearch/issues/17213) on 1 or few nodes in the hot tier, a [new warm node receiving more than its fair share of shards](https://github.com/elastic/elasticsearch/issues/51422) from high-shard-count indices, etc..).

_Many_ users have hit this (I just had another user hit this yesterday). Unfortunately, they tend to find out when their clusters are already performing poorly with specific nodes getting hammered with requests.  Until the two high hanging fruit issues are addressed in the future, it can be a nice addition to our documentation to provide the common use case of `total_shards_per_node` to prevent hotspots in the cluster (certainly, also indicating the tradeoff).  

2 potential places to provide this guidance:
- https://www.elastic.co/guide/en/elasticsearch/reference/7.9/allocation-total-shards.html
- Add a new page under "Recipes": https://www.elastic.co/guide/en/elasticsearch/reference/7.9/recipes.html 

If we decide to make this doc improvement, we may want to do this around the time we [enhance ILM](https://github.com/elastic/elasticsearch/issues/44070) to allow updating of `total_shards_per_node` in ILM phases.

### PR DESCRIPTION
**Pull request by jrodewig**
Revises the current 'How to avoid oversharding' docs to incorporate
information from our [shard sizing blog post][0].

Changes:

* Streamlines introduction
* Adds "Things to remember" section to describe how shards work
* Adds "Guidelines" section based on blog tips
* Creates a "Fix an oversharded cluster" section

[0]: https://www.elastic.co/blog/how-many-shards-should-i-have-in-my-elasticsearch-cluster

Closes #52541, closes #61306

### Preview
https://elasticsearch_61942.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/size-your-shards.html

### CODE REVIEW

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+==== Each shard has overhead
+
+Every shard uses file handles, memory, and CPU resources. In most cases, a small
+set of large shards uses fewer resources than a large set of small shards.
+
+Segments play a large role in a shard's resource usage. Most shards contain
+several segments, which store its index data. For faster searches, segment
+metadata is kept in <<heap-size,heap memory>>. As a shard grows, its segments
+are <<index-modules-merge,merged>> into fewer, larger segments. As the number of
+segments decreases, less metadata needs to be kept in heap memory.
```
Comment (jrodewig) [inline docs/reference/how-to/size-your-shards.asciidoc:64]: Note for reviewers:

Is this still the case with our recent heap improvements in 7.7?
https://www.elastic.co/guide/en/elasticsearch/reference/7.7/release-highlights.html#_significant_reduction_of_heap_usage_of_segments

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+system resources.
+
+[discrete]
+[[shard-count-recommendation]]
+==== Aim for 20 shards or fewer per GB of heap space
+
+The number of shards a node can hold is proportional to the node's
+<<heap-size,heap space>>. For example, a node with a 30GB heap should have a
+maximum of 600 shards. If you find your nodes nearing more than 20 shards per
+GB, consider adding another node.
```
Comment (jrodewig) [inline docs/reference/how-to/size-your-shards.asciidoc:141]: Note for reviewers:

Note sure if this is still our guideline after heap improvements:
https://www.elastic.co/guide/en/elasticsearch/reference/7.7/release-highlights.html#_significant_reduction_of_heap_usage_of_segments

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+Every shard uses file handles, memory, and CPU resources. In most cases, a small
+set of large shards uses fewer resources than many small shards.
+
+Segments play a big role in a shard's resource usage. Most shards contain
+several segments, which store its index data. {es} keeps segment metadata in
+<<heap-size,heap memory>> so it can be quickly retrieved for searches. As a
+shard grows, its segments are <<index-modules-merge,merged>> into fewer, larger
+segments. This decreases the number of segments, which means less metadata is
+kept in heap memory.
```
Comment (jrodewig) [inline docs/reference/how-to/size-your-shards.asciidoc:65]: Note for reviewers:

I'm not sure how much we want to emphasize this now that we've reduced heap usage:
https://www.elastic.co/guide/en/elasticsearch/reference/7.7/release-highlights.html#_significant_reduction_of_heap_usage_of_segments

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+[discrete]
+[[shard-count-recommendation]]
+==== Aim for 20 shards or fewer per GB of heap space
+
+The number of shards a node can hold is proportional to the node's
+<<heap-size,heap space>>. For example, a node with a 30GB heap should have at
+most 600 shards. The further below this limit you can keep your nodes, the
+better. If you find your nodes exceeding more than 20 shards per GB, consider
+adding another node. You can use the cat shards API to check the number of
+shards per node.
```
Comment (jrodewig) [inline docs/reference/how-to/size-your-shards.asciidoc:147]: Note for reviewers:

Not sure if this guideline still hold now that heap usage is reduced:
https://www.elastic.co/guide/en/elasticsearch/reference/7.7/release-highlights.html#_significant_reduction_of_heap_usage_of_segments

Any update guidance is appreciated!

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+{
+  "source": {
+    "index": "my-oversharded-index-000001"
+  },
+  "dest": {
+    "index": "my-new-index-000001"
+  }
+}
+----
+// TEST[s/^/PUT my-oversharded-index-000001\n/]
```
Comment (jpountz) [inline docs/reference/how-to/size-your-shards.asciidoc:289]: This is a slower, more resource-intensive version of shrinking, so I wonder if we should mention it at all.

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+== How to size your shards
+++++
+<titleabbrev>Size your shards</titleabbrev>
+++++
+
+To protect against hardware failure and increase capacity, {es} stores copies of
+an index’s data across multiple shards on multiple nodes. The number and size of
+these shards can have a significant impact on your cluster's health. One common
+problem is _oversharding_, a situation in which a cluster with a large number of
+shards has degraded performance.
```
[the remark this answers is not in the document]
Comment (jrodewig) [reply docs/reference/how-to/size-your-shards.asciidoc]: Thanks for the feedback. Updated with ce5f86c4a9f483461c39672bf03a2e26385a8036.

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+When possible, delete entire indices instead. {es} can immediately remove
+deleted indices directly from the file system and free up resources.
+
+[discrete]
+[[use-ds-ilm-for-time-series]]
+==== For time series data, use data streams and {ilm-init}
+
+<<data-streams,Data streams>> let you store time series data across multiple,
+time-based backing indices. You can use <<index-lifecycle-management,{ilm}
+({ilm-init})>> lifecycle policies to automatically manage these backing indices.
```
[the remark this answers is not in the document]
Comment (jrodewig) [reply docs/reference/how-to/size-your-shards.asciidoc]: Removed with ce5f86c4a9f483461c39672bf03a2e26385a8036.

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+
+[discrete]
+[[avoid-node-hotspots]]
+==== Avoid node hotspots
+
+If you use a hot-warm architecture, you can use
+<<shard-allocation-awareness,shard allocation awareness>> to assign shards to
+nodes based on physical hardware. However, if too many shards are allocated to a
+specific node, the node can become a hotspot. For example, if a single node
+contains all the primary shards for an index with a high indexing volume, the
```
[the remark this answers is not in the document]
Comment (jrodewig) [reply docs/reference/how-to/size-your-shards.asciidoc]: Updated the wording with 6a2277bdb70.

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+setting to explicitly limit the number of shards on a single node. You can
+configure `index.routing.allocation.total_shards_per_node` using the
+<<indices-update-settings,update index settings API>>.
+
+[source,console]
+--------------------------------------------------
+PUT /my-index-000001/_settings
+{
+  "index" : {
+    "routing.allocation.total_shards_per_node" : 400
```
[the remark this answers is not in the document]
Comment (jrodewig) [reply docs/reference/how-to/size-your-shards.asciidoc]: Good catch. Updated with ce5f86c4a9f483461c39672bf03a2e26385a8036

File: docs/reference/how-to/size-your-shards.asciidoc (hunk 5 above)
Comment (jrodewig) [reply docs/reference/how-to/size-your-shards.asciidoc]: +1. Removed with ce5f86c4a9f483461c39672bf03a2e26385a8036.

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+[[shard-auto-balance]]
+==== {es} automatically balances shards across nodes
+
+By default, {es} attempts to spread an index's shards across as many nodes as
+possible. When a node fails or you add a new node, {es} automatically rebalances
+shards across the remaining nodes. However, this behavior isn't always wanted.
+If you use hot-warm architecture, you likely want shards for older indices on
+nodes with less expensive hardware and vice versa. You can use
+<<shard-allocation-awareness,shard allocation awareness>> to ensure shards for
+particular indices only move to specified nodes.
```
[the remark this answers is not in the document]
Comment (jrodewig) [reply docs/reference/how-to/size-your-shards.asciidoc]: Simplified this a bit with 1f856d20c1f8be44d3ce8ca4bd2586a371b6e07c. Would still like any feedback from @dakrone  or @debadair. I'll also add a cross-link when the data tier docs are live.

**Review by jpountz — APPROVED** [review]

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+{kibana-ref}/elasticsearch-metrics.html[{es} monitoring tools] to track your
+cluster's stability and performance.
+
+The following sections provide some reminders and guidelines you should consider
+when designing your sharding strategy. If your cluster has shard-related
+problems, see <<fix-an-oversharded-cluster>>.
+
+[discrete]
+[[shard-size-reminders]]
+=== Things to remember
```
Comment (debadair) [inline docs/reference/how-to/size-your-shards.asciidoc:40]: "Things to remember" seems a bit vague. Maybe "Sizing considerations" and then use "things" in the intro sentence?

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+==== {es} automatically balances shards within a data tier
+
+A cluster's nodes are grouped into data tiers. Within each tier, {es}
+attempts to spread an index's shards across as many nodes as possible. When you
+add a new node or a node fails, {es} automatically rebalances the index's shards
+across the tier's remaining nodes.
+
+[discrete]
+[[shard-size-guidelines]]
+=== Guidelines
```
Comment (debadair) [inline docs/reference/how-to/size-your-shards.asciidoc:78]: These feel more like best practices than guidelines.

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+Instead, {es} marks the document as deleted on each related shard. The marked
+document will continue to use resources until it's removed during a periodic
+<<index-modules-merge,segment merge>>.
+
+When possible, delete entire indices instead. {es} can immediately remove
+deleted indices directly from the file system and free up resources.
+
+[discrete]
+[[use-ds-ilm-for-time-series]]
+==== For time series data, use data streams and {ilm-init}
```
Comment (debadair) [inline docs/reference/how-to/size-your-shards.asciidoc:97]: I'd lead with the imperative, "Use data streams..."

File: docs/reference/how-to/size-your-shards.asciidoc [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+=== Fix an oversharded cluster
+
+If your cluster is experiencing stability issues due to oversharded indices,
+you can use one or more of the following methods to fix them.
+
+[discrete]
+[[reindex-indices-from-shorter-periods-into-longer-periods]]
+==== Create time-based indices that cover longer periods
+
+For time series data, you can create indices that cover longer time intervals.
```
Comment (debadair) [inline docs/reference/how-to/size-your-shards.asciidoc:198]: Seems like this should recommend using criteria other than the time period.

**Review by debadair — APPROVED** [review]
Left a couple minor editorial suggestions, but this is great!

**Comment by jrodewig** [comment]
Thanks for the feedback @debadair!

**Review by debadair — APPROVED** [review]

**Comment by jrodewig** [comment]
Will go ahead and merge. I'll plan to add the data tier xrefs when those docs are released.

**Pull request merged**